source-braintree-native/source_braintree_native/api.py

```diff
@@ -108,36 +132,134 @@ async def snapshot_resources(
         yield FullRefreshResource.model_validate(_braintree_object_to_dict(object))
 
 
+async def _fetch_transaction_ids_for_single_field(
+    field: str,
+    braintree_gateway: BraintreeGateway,
+    start: datetime,
+    end: datetime,
+    log: Logger,
+) -> list[str]:
+
+    search_criteria: Search.RangeNodeBuilder = getattr(TransactionSearch, field)
+
+    collection: ResourceCollection = await asyncio.to_thread(
+        braintree_gateway.transaction.search,
+        search_criteria.between(start, end),
+    )
+
+    if collection.maximum_size >= TRANSACTION_SEARCH_LIMIT:
+        raise RuntimeError(_search_limit_error_message(collection.maximum_size, "transactions"))
+
+    return collection.ids
+
+
+async def _fetch_unique_updated_transaction_ids(
+    braintree_gateway: BraintreeGateway,
+    start: datetime,
+    end: datetime,
+    log: Logger,
+) -> list[str]:
+    async with asyncio.TaskGroup() as tg:
+        tasks = [tg.create_task(_fetch_transaction_ids_for_single_field(
+            field,
+            braintree_gateway,
+            start,
+            end,
+            log
+        )) for field in TRANSACTION_SEARCH_FIELDS]
+
+    id_set: set[str] = set()
+    for task in tasks:
+        for id in task.result():
+            id_set.add(id)
+
+    return list(id_set)
+
+
+async def _fetch_transaction_batch(
+        braintree_gateway: BraintreeGateway,
+        ids: list[str],
+        queue: asyncio.Queue,
+        log: Logger,
+):
+    collection: ResourceCollection = await asyncio.to_thread(
+        braintree_gateway.transaction.search,
+        TransactionSearch.ids.in_list(ids),
+    )
+
+    async for object in _async_iterator_wrapper(collection):
+        doc = IncrementalResource.model_validate(_braintree_object_to_dict(object))
+        await queue.put(doc)
+
+
 async def fetch_transactions(
         braintree_gateway: BraintreeGateway,
         window_size: int,
         log: Logger,
         log_cursor: LogCursor,
 ) -> AsyncGenerator[IncrementalResource | LogCursor, None]:
     assert isinstance(log_cursor, datetime)
-    most_recent_created_at = log_cursor
     window_end = log_cursor + timedelta(hours=window_size)
     end = min(window_end, datetime.now(tz=UTC))
 
+    ids = await _fetch_unique_updated_transaction_ids(
+        braintree_gateway,
+        log_cursor,
+        end,
+        log,
+    )
+
+    queue = asyncio.Queue(maxsize=10)
+
+    async with asyncio.TaskGroup() as tg:
```

Comment:
You could probably use `asyncio.gather` here as well.

A possible consideration is memory usage if there are a lot of batches which all get results concurrently before their documents can be emitted. I doubt this will be a problem in practice since the speed at emitting documents should be quite quick in comparison to fetching batches.

Reply:
I'll refactor to the `asyncio.gather` pattern you linked in your other comment.

I agree about memory usage. Theoretically with the current batch size of `TRANSACTION_SEARCH_LIMIT`, there could be at max 9 batches (the same as the the number of transaction search fields) if all of them return unique transaction IDs within the time window. In practice, there's a lot of overlap between some of these (like `created_at` and `submitted_for_settlement_at`), and searches for the non-`created_at` fields usually return fewer IDs than `created_at` searches. I could add some handling to only allow X batches to fetch results at a time, but I suggest we could wait & see if that's actually needed.

Reply:
I was able to refactor this one, but I had to omit the `await`. Otherwise, the connector would get stuck putting docs onto the queue but never taking them off.

Reply:
The change in `_fetch_unique_updated_transaction_ids` looks good 👍 

For `fetch_transactions` I see how the same approach won't work, since we specifically _don't_ want to keep every single full record in memory.

I'm not 100% sure what you've got in `fetch_transactions` right now will work. I think the gather has to be await'd at some point for it to be assured to complete. 

As an alternative to using a queue, take a look at `asyncio.as_completed` - I think that would do what we want, which is to eagerly process the results of each task as it becomes available.

Reply:
Based off our discussion, I made `fetch_transactions` fetch batches in sequence instead of in parallel. It'll probably be uncommon to have more than 1-2 batches while we use a batch size of 50,000. We can look into reducing batch size & processing the batches in parallel at a later point.